core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -144,23 +148,56 @@ public long acquireExecutionMemory(long required, MemoryConsumer consumer) {
       // spilling, avoid to have too many spilled files.
       if (got < required) {
         // Call spill() on other consumers to release memory
+        // Sort the consumers according their memory usage. So we avoid spilling the same consumer
+        // which is just spilled in last few times and re-spilling on it will produce many small
+        // spill files.
+        TreeMap<Long, List<MemoryConsumer>> sortedConsumers =
+          new TreeMap<Long, List<MemoryConsumer>>();
         for (MemoryConsumer c: consumers) {
           if (c != consumer && c.getUsed() > 0 && c.getMode() == mode) {
-            try {
-              long released = c.spill(required - got, consumer);
-              if (released > 0) {
-                logger.debug("Task {} released {} from {} for {}", taskAttemptId,
-                  Utils.bytesToString(released), c, consumer);
-                got += memoryManager.acquireExecutionMemory(required - got, taskAttemptId, mode);
-                if (got >= required) {
-                  break;
-                }
+            long key = c.getUsed();
+            List<MemoryConsumer> list = null;
+            if (sortedConsumers.containsKey(key)) {
+              list = sortedConsumers.get(key);
+            } else {
+              list = new ArrayList<MemoryConsumer>(1);
+            }
+            list.add(c);
+            sortedConsumers.put(key, list);
+          }
+        }
+        while (true) {
```

Comment:
Use `while (!sortedConsumers.isEmpty)` - and remove `if (currentEntry == null)` below (line 179) since it cant fail.